dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.42/V2_42_26__Add_not_null_constraint_to_trackedentity.sql

```diff
@@ -0,0 +1,39 @@
+DO $$
+DECLARE
+results RECORD;
+inconsistent_records_count INTEGER;
+BEGIN
+SELECT COUNT(*) INTO inconsistent_records_count FROM trackedentity WHERE trackedentitytypeid IS NULL;
+
+IF inconsistent_records_count > 0 THEN
+    RAISE NOTICE 'Inconsistencies found: %', inconsistent_records_count;
+
+    FOR results IN
+    SELECT programid, trackedentityid FROM enrollment WHERE trackedentityid IN (
+        SELECT trackedentityid
+        FROM trackedentity
+        WHERE trackedentitytypeid IS NULL
+    )
+    LOOP
+    UPDATE trackedentity
+    SET trackedentitytypeid = (
+        SELECT trackedentitytypeid
+        FROM program
+        WHERE programid = results.programid
+    )
+    WHERE trackedentityid = results.trackedentityid;
+
+    RAISE NOTICE 'Updated trackedentity: programid = %, trackedentityid = %', results.programid, results.trackedentityid;
+    END LOOP;
+    SELECT COUNT(*) INTO inconsistent_records_count FROM trackedentity WHERE trackedentitytypeid IS NULL;
+
+    IF inconsistent_records_count = 0 THEN
+        ALTER TABLE trackedentity ALTER COLUMN trackedentitytypeid SET NOT NULL;
+    ELSE
```

Comment:
This logic is a little bit redundant.
You can apply the constraint and if it fails then you raise the exception.